### ISSUE TITLE
Smoothstep breaks GLSL compilation when multiple type signature nodes are used

### ISSUE THREAD

**Issue by HardCoreCodin** [AcademySoftwareFoundation/MaterialX#1978]
When multiple signature variants of smoothstep participate in a graph, the viewport render goes blank and this shows in the output:
```
Error in compiling fragment shader:
0(493) : error C1013: function "mx_smoothstep_float" is already defined at 0(483)
```

This is likely because the function `mx_smoothstep_float` exists in multiple headers that get included in such cases. 

See attached example file.
[smoothstep_failure_mtlx.txt](https://github.com/user-attachments/files/16658780/smoothstep_failure_mtlx.txt)

**Comment by kwokcb** [comment]
There are includes which I assume not being checked to see if that file has already been included in the non-float code implementations
```
#include "mx_smoothstep_float.glsl"
```
This breaks GLSL and Metal at least.

**Comment by jstone-lucasfilm** [comment]
Good catch, @HardCoreCodin, and this sounds like a bug in our GLSL shader generation that should be fixed.

**Comment by ld-kerley** [comment]
I think the fix is probably a pretty easy `#ifndef/#define` trick in the right place - I can investigate and put up a PR soon

**Comment by kwokcb** [comment]
@ld-kerley , not sure if this is the place to look but I believe `#include`s are parsed and code is embedded -- but it should not try to include twice based on `ShaderStage::addInclude()`.

**Comment by ld-kerley** [comment]
In this case the `#include` statement is in a file that is already being included - a quick local test and adding a standard include guard around the actual source does appear to work in both GLSL and MSL.

**Comment by kwokcb** [comment]
I’ll just throw out one more option. Turn the vector versions of this into graphs that use the float version node instances — hence no include dependence exists. Would have to think some more on the other options, but adding the common include seems like a good idea if it is added on demand.

**Comment by ld-kerley** [comment]
I like the nodegraph implementation idea, though if we applied this to all destination languages, then I think we would be constructing less optimal OSL code.

```
  <implementation name="IM_smoothstep_vector3_genosl" nodedef="ND_smoothstep_vector3" target="genosl" sourcecode="smoothstep({{low}}, {{high}}, {{in}})" />
```

OSL currently provides a `smoothstep()` that can take a `vector` type that will allow for more aggressive OSL optimizer optimizations. 

I guess that raises a hole in my MaterialX understanding.  If a `nodegraph` implementation exists, alongside a target specific `implementation` as above, what happens?  I guess the implementation that is more specific would ideally be selected, if so then we could safely deploy the nodegraph idea (which I think in general is how we should try and build all nodes to make them more maintainable), without any loss of OSL performance.

**Comment by ld-kerley** [comment]
Interestingly we already have "some" `smoothstep` nodegraph implementations, so I think that casts an even stronger vote for that solution.

**Comment by dbsmythe** [comment]
Yes, according to the Spec:

> It is allowable for there to be both a <nodegraph> and an <implementation> for the same nodedef target/version, with the <implementation> generally prevailing in order to allow for optimized native-code node implementations, although ultimately it would be up to the host application to determine which implementation to actually use.

(The above sentence is in the paragraph right **above** the https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/documents/Specification/MaterialX.Specification.md#example-custom-node-defined-by-a-nodegraph section)

**Comment by HardCoreCodin** [comment]
There may be strong reasons in the near term to also have a custom `smoothstep` function, even for float, for all back-end targets, including OSL. And they would all likely need to change soon - see this other discussion on the [OSL docs issue](https://github.com/AcademySoftwareFoundation/OpenShadingLanguage/pull/1851) (it would also related to this GLSL implementation from the MaterialX's perspective).

**Comment by kwokcb** [comment]
@ld-kerley for your implementation matching question:
-  The target is sent over [when searching for the implementation ](https://github.com/AcademySoftwareFoundation/MaterialX/blob/153a803c46181319fd782ef8426ff58a2e885d82/source/MaterialXGenShader/ShaderGenerator.cpp#L282) so it should [search for the anything that matches the `target` first ](https://github.com/AcademySoftwareFoundation/MaterialX/blob/153a803c46181319fd782ef8426ff58a2e885d82/source/MaterialXCore/Definition.cpp#L76)before checking non-target specific implementations.
- The `target` value comes from the code generator.

**Comment by jstone-lucasfilm** [comment]
Thanks for this report, @HardCoreCodin, and to @ld-kerley for the fix in #1982!
